cpp/es2.1/dateEdoardo.cpp

```diff
@@ -0,0 +1,153 @@
+#include<iostream>
+
+using namespace std;
+
+class Date {
+    private:
+        int day, month, year;
+
+    public:
+        Date() {day=1; month=1; year=1970;}
+        Date(int day, int month, int year) {this->day=day; this->month=month; this->year=year;}
+        int getDay() {return day;}
+        int getMonth() {return month;}
+        int getYear() {return year;}
+        friend ostream& operator<<(ostream &out, Date &pt);
+        friend bool operator==(Date &first, Date &second);
+        friend void addDays(int n, Date d);
+        friend void addMonth(int n, Date d);
+        friend void addYear(int n, Date d);
+        friend void adjust(Date d);
+};
+ostream& operator<<(ostream &out, Date &pt){
+    out<<pt.day<<"-"<<pt.month<<"-"<<pt.year<<endl;
+    return out;
+}
+bool operator==(Date &first, Date &second){
+    return (first.day==second.day)&&(first.month==second.month)&&(first.year==second.year);
+}
+void addDays(int n, Date d){
+    d.day+=n;
+    adjust(d);
+}
+void addMonth(int n, Date d){
+    d.month+=n;
+    adjust(d);
+}
+void addYear(int n, Date d){
+    d.year+=n;
+    adjust(d);
+}
+void adjust(Date d){
+    bool esci;
+    do{
+    esci=true;
+    int days;
+    if(d.month==1||d.month==3||d.month==5||d.month==7||d.month==8||d.month==10||d.month==12)
+        days=31;
+    else if(d.month==4||d.month==6||d.month==9||d.month==11)
+        days=30;
+    else
+        if(d.year%4==0)
+            days=29;
+        else
+            days=28;
+
+    if(d.day>days){
+        esci=false;
+        d.day-=days;
+        d.month++;
+    }
+
+    if(d.month>12){
+        esci=false;
+        d.month-=12;
+        d.year++;
+    }
+    }
+    while(!esci);
+}
+int main(){
+    cout<<"1) inserire data\n2) data default (1 gen 1970)\n";
+    int scelta;
+    Date date;
+    cin>>scelta;
+    if(scelta==1){
+        int anno, mese, giorno;
+        cout<<"anno?\n";
+        cin>>anno;
+        cout<<"mese?\n";
+        cin>>mese;
+        cout<<"giorno?\n";
+        cin>>giorno;
+        Date date {giorno, mese, anno};
+    }
```

Comment:
In questo esercizio stiamo implementando un nuovo tipo `Date`, per il quale è importante avere molti test per evitare regressioni durante lo sviluppo.  Qua stai usando un metodo di test interattivo, che però è molto lento.
Prova a implementare una serie di test di unità in questo modo:
```c++
// apri un blocco per poter definire variabili più volte
{
  // test addYears(1)
  Date d(2023, 6, 22);
  d.addYears(1)
  if (d == Date(2024, 6, 22))
    std::cout << "[OK] addYears(1)";
  else
    std::cout << "[FAIL] addYears(1)";
}
{
  // test addYears() maggiore di 1
  Date d(2023, 6, 22);
  d.addYears(4)
  if (d == Date(2027, 6, 22))
    std::cout << "[OK] addYears(4)";
  else
    std::cout << "[FAIL] addYears(4)";
}
{
  // test addYears() -1
  Date d(2023, 6, 22);
  d.addYears(-1)
  if (d == Date(2022, 6, 22))
    std::cout << "[OK] addYears(-1)";
  else
    std::cout << "[FAIL] addYears(-1)";
}
```
E così via. Per prima cosa devi testare il costruttore di default e il costruttore con parametri, verifica che days, months e years siano quelli che hai messo. Poi testa l'operatore di uguaglianza `operator==`.
DOPO puoi passare a scrivere gli altri test, e cioè `addMonths(<numero positivo>)`, `addMonths(<numero negativo>)`, `addDays(<numero positivo>)` etc.